### PR TITLE
[ISSUE-696] logging improvements

### LINKED ISSUE #696: Bare-metal CSI Logging Improvements
**Is your feature request related to a problem? Please describe.**
Need to improve logging in Baremetal CSI Driver and Operator.

**Describe the solution you'd like**
KubeClient prints raw custom resource data. INFO log level should be changed to DEBUG:
```
Nov 25 06:43:40.695 [INFO] [KubeClient] [UpdateCR] [9a952a8d-1a31-4e7d-b36c-92afea0131d4] Updating CR LogicalVolumeGroup, &{{LogicalVolumeGroup csi-baremetal.dell.com/v1} {9a952a8d-1a31-4e7d-b36c-92afea0131d4    db861ecf-2520-4621-bf57-7aea59da0ae3 8429440 25 2021-11-10 13:05:25 +0000 UTC <nil> <nil> map[app:csi-baremetal app.kubernetes.io/name:csi-baremetal] map[lvg/free-space:324966285312] [] []  [{node Update csi-baremetal.dell.com/v1 2021-11-22 14:45:42 +0000 UTC FieldsV1 &FieldsV1{Raw:*[123 34 102 58 109 101 116 97 100 97 116 97 34 58 123 34 102 58 97 110 110 111 116 97 116 105 111 110 115 34 58 123 34 46 34 58 123 125 44 34 102 58 108 118 103 47 102 114 101 101 45 115 112 97 99 101 34 58 123 125 125 44 34 102 58 108 97 98 101 108 115 34 58 123 34 46 34 58 123 125 44 34 102 58 97 112 112 34 58 123 125 44 34 102 58 97 112 112 46 107 117 98 101 114 110 101 116 101 115 46 105 111 47 110 97 109 101 34 58 123 125 125 125 44 34 102 58 115 112 101 99 34 58 123 34 46 34 58 123 125 44 34 102 58 72 101 97 108 116 104 34 58 123 125 44 34 102 58 76 111 99 97 116 105 111 110 115 34 58 123 125 44 34 102 58 78 97 109 101 34 58 123 125 44 34 102 58 78 111 100 101 34 58 123 125 44 34 102 58 83 105 122 101 34 58 123 125 44 34 102 58 83 116 97 116 117 115 34 58 123 125 125 125],}} {controller Update csi-baremetal.dell.com/v1 2021-11-22 23:25:23 +0000 UTC FieldsV1 &FieldsV1{Raw:*[123 34 102 58 115 112 101 99 34 58 123 34 102 58 86 111 108 117 109 101 82 101 102 115 34 58 123 125 125 125],}}]} {root-vg 615c78ef-8041-4a59-9e3c-4cd287de0597 [bff4755c-19c3-4b3b-8fdb-5ba0677d28a4] 324966285312 [lv_root lv_swap lv_var] CREATED GOOD {} [] 0}}
```
Use %+v for structs. Not sure, what is {} or [] here
```
Nov 29 08:15:17.290 [INFO] [VolumeOperationsImpl] [CreateVolume] [pvc-acae3224-c75b-43dc-95c2-3ce11d3f1608] Creating volume {pvc-acae3224-c75b-43dc-95c2-3ce11d3f1608   ANY a2a58c0f-6a6f-469b-83dc-ea7c79dbd3cc [] 62914560 FS xfs     false {} [] 0}
```
Could we print func and file in error messages? Example:
```
time="2022-01-07T12:25:02Z" level=error msg="method failed" func=github.com/piraeusdatastore/linstor-csi/pkg/driver.Driver.Run.func1 file="/usr/local/go/linstor-csi/pkg/driver/driver.go:1150" error="rpc error: code = Internal desc = NodePublishVolume failed for pvc-ee1edd44-f48d-4a46-84ee-9434c2859581: failed to set source device readwrite: exit status 1"
```
Operator:
```
si-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:38Z" level=warning msg="Invalid number of csi deployments at RoleBinding watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:40Z" level=warning msg="Invalid number of csi deployments at Role watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
csi-baremetal-operator-597774cbfc-trsfw/manager@kind..ker4: time="2022-01-11T03:10:40Z" level=warning msg="Invalid number of csi deployments at Role watcher, number: '0', expected: '1'" component=DeploymentReconciler module=controllers
```

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Aber4Nod**
### Resolves #696 

Improving logging in Baremetal CSI Driver and Operator.

## PR checklist
- [ ] Add link to the issue
- [ ] Choose Project
- [ ] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
_Provide test details_

### CODE REVIEW

File: pkg/base/k8s/kubeclient.go [hunk 1]
```diff
@@ -104,7 +104,7 @@ func (k *KubeClient) CreateCR(ctx context.Context, name string, obj runtime.Obje
 		"requestUUID": requestUUID.(string),
 	})
 	crKind := obj.GetObjectKind().GroupVersionKind().Kind
-	ll.Infof("Creating CR %s: %v", crKind, obj)
```
Comment (mishoyama) [inline pkg/base/k8s/kubeclient.go:107]: Is it possible to not print raw data at all, but `spec` field with INFO level?

File: pkg/base/logger/runtime_formatter.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	data := logrus.Fields{}
+	if f.MaxLevel >= entry.Level {
+		function, file, line := f.getCurrentPosition(entry)
+		packageEnd := strings.LastIndex(function, ".")
+		functionName := function[packageEnd+1:]
+
+		data[FunctionKey] = functionName
+		data[FileKey] = file + ":" + line
+	}
+	for k, v := range entry.Data {
```
Comment (mishoyama) [inline pkg/base/logger/runtime_formatter.go:39]: Do we really want to have this overhead for logging?

Comment (Aber4Nod) [reply pkg/base/logger/runtime_formatter.go:39]: 1. This format is node only for error level, not for logging
2. This was such requirement

File: pkg/base/k8s/kubeclient.go (hunk 1 above)
Comment (Aber4Nod) [reply pkg/base/k8s/kubeclient.go:107]: The requirement was to print debug with raw. The requirement was changed?

Comment (safronovD) [reply pkg/base/k8s/kubeclient.go:107]: + labels and annotations, if possible

Comment (Aber4Nod) [reply pkg/base/k8s/kubeclient.go:107]: Can you please write the requirement clearly - what should be done here? Because currently there are 3 options about what is needed to do with this log

Comment (Aber4Nod) [reply pkg/base/k8s/kubeclient.go:107]: Pushed changed with logging CSI Objects fields - labels, annotations and spec at info level

**Review by safronovD — APPROVED** [review]

**Comment by rkiselenko** [comment]
What the point to support all kind of objects github.com/dell/csi-baremetal/pkg/base/logger/objects as separated package ? 
All those objects formatted with the same string `Labels: %+v, Annotations: %+v, Spec: %+v`. 
AFAIK we can have only one function, something like this:

```golang
func (l *objectLogger) Log(object runtime.Object) string {
  // some code here
  formattedString := ""
  switch {
  case gvk.Kind == acKind:
    formattedString = fmt.Sprintf("Labels: %+v, Annotations: %+v, Spec: %+v", object.ObjectMeta.Labels, object.ObjectMeta.Annotations, object.Spec)
    // some code goes here
  default:
    formattedString = fmt.Sprintf("%+v", object)
   }
   return formattedString
}
```

What do you think?

File: pkg/base/logger/runtime_formatter.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+// FileKey holds the file field
+const FileKey = "file"
+
+const (
+	logrusStackJump          = 4
+	logrusFieldlessStackJump = 6
+)
+
+// RuntimeFormatter decorates log entries with function name and package name (optional) and line number (optional)
```
Comment (rkiselenko) [inline pkg/base/logger/runtime_formatter.go:22]: The package name should be provided as [fields](https://pkg.go.dev/github.com/sirupsen/logrus#readme-fields) 
Because of `json` log formatter.

**Comment by Aber4Nod** [comment]
> What the point to support all kind of objects github.com/dell/csi-baremetal/pkg/base/logger/objects as separated package ? All those objects formatted with the same string `Labels: %+v, Annotations: %+v, Spec: %+v`. AFAIK we can have only one function, something like this:
> 
> ```go
> func (l *objectLogger) Log(object runtime.Object) string {
>   // some code here
>   formattedString := ""
>   switch {
>   case gvk.Kind == acKind:
>     formattedString = fmt.Sprintf("Labels: %+v, Annotations: %+v, Spec: %+v", object.ObjectMeta.Labels, object.ObjectMeta.Annotations, object.Spec)
>     // some code goes here
>   default:
>     formattedString = fmt.Sprintf("%+v", object)
>    }
>    return formattedString
> }
> ```
> 
> What do you think?

Because it's flexible (if we need custom logic in the future over logging per object), easy to scale, easy to understand. It's backed to design pattern and hence is a common practice - not just put everything in one superfunction

**Comment by rkiselenko** [comment]
> > What the point to support all kind of objects github.com/dell/csi-baremetal/pkg/base/logger/objects as separated package ? All those objects formatted with the same string `Labels: %+v, Annotations: %+v, Spec: %+v`. AFAIK we can have only one function, something like this:
> > ```go
> > func (l *objectLogger) Log(object runtime.Object) string {
> >   // some code here
> >   formattedString := ""
> >   switch {
> >   case gvk.Kind == acKind:
> >     formattedString = fmt.Sprintf("Labels: %+v, Annotations: %+v, Spec: %+v", object.ObjectMeta.Labels, object.ObjectMeta.Annotations, object.Spec)
> >     // some code goes here
> >   default:
> >     formattedString = fmt.Sprintf("%+v", object)
> >    }
> >    return formattedString
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What do you think?
> 
> Because it's flexible (if we need custom logic in the future over logging per object), easy to scale, easy to understand. It's backed to design patterns and hence is a common practice - not just put everything in one superfunction

But you still have a way to write custom logic (logging per object) without a lot of ditto packages. It's better to write less code.
In my opinion this code is overengineening.

File: pkg/base/logger/runtime_formatter.go (hunk 3 above)
Comment (Aber4Nod) [reply pkg/base/logger/runtime_formatter.go:22]: Can you please write the comment in details

**Comment by Aber4Nod** [comment]
> > > What the point to support all kind of objects github.com/dell/csi-baremetal/pkg/base/logger/objects as separated package ? All those objects formatted with the same string `Labels: %+v, Annotations: %+v, Spec: %+v`. AFAIK we can have only one function, something like this:
> > > ```go
> > > func (l *objectLogger) Log(object runtime.Object) string {
> > >   // some code here
> > >   formattedString := ""
> > >   switch {
> > >   case gvk.Kind == acKind:
> > >     formattedString = fmt.Sprintf("Labels: %+v, Annotations: %+v, Spec: %+v", object.ObjectMeta.Labels, object.ObjectMeta.Annotations, object.Spec)
> > >     // some code goes here
> > >   default:
> > >     formattedString = fmt.Sprintf("%+v", object)
> > >    }
> > >    return formattedString
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > What do you think?
> > 
> > 
> > Because it's flexible (if we need custom logic in the future over logging per object), easy to scale, easy to understand. It's backed to design patterns and hence is a common practice - not just put everything in one superfunction
> 
> But you still have a way to write custom logic (logging per object) without a lot of ditto packages. It's better to write less code. In my opinion this code is overengineening.

In my opinion and experience of writing/supporting/driving large scale systems in one repo it is not. So, I disegree with you in this point.

File: pkg/base/logger/runtime_formatter.go (hunk 3 above)
Comment (rkiselenko) [reply pkg/base/logger/runtime_formatter.go:22]: The services like kibana parse json logs in the elastic index, to have package name parsed as expected it should be field (json key\value) otherwise it's hard to parse package name from plain string.

Comment (Aber4Nod) [reply pkg/base/logger/runtime_formatter.go:22]: And what't wrong with this package name?

Comment (rkiselenko) [reply pkg/base/logger/runtime_formatter.go:22]: I was thinking it's for log entry but it for package function.

File: pkg/base/logger/runtime_formatter.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	if len(entry.Data) == 0 {
+		skip = logrusFieldlessStackJump
+	}
+start:
+	pc, file, line, _ := runtime.Caller(skip)
+	lineNumber := fmt.Sprintf("%d", line)
+	function := runtime.FuncForPC(pc).Name()
+	if strings.LastIndex(function, "sirupsen/logrus.") != -1 {
+		skip++
+		goto start
```
[the remark this answers is not in the document]
Comment (Aber4Nod) [reply pkg/base/logger/runtime_formatter.go:58]: Goto is done to 6 lines higher. So, a bit of side information here:
1. Goto is a very efficient way of writing code - it's just a simple instruction, that is one of easiest from the productivity of view at all.
2. It's a common way of writing code, especially in C language and is used particularly in every well-known system/application.
3. It's a good practice (and yes - forget what teachers says at the univercity :)) to use this mechanism in application - BUT if you use this carefully.

So, it can be used here, because code is very easy to be read, there are no performance overhead at all,

[the remark this answers is not in the document]
Comment (rkiselenko) [reply pkg/base/logger/runtime_formatter.go:58]: 1. indeed
2. C != Go
3. indeed

This code can be implemented without `goto`.

Using `runtime` package leads to performance issue.

**Review by rkiselenko — CHANGES_REQUESTED** [review]

**Review by mishoyama — APPROVED** [review]

File: pkg/base/logger/runtime_formatter.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Aber4Nod) [reply pkg/base/logger/runtime_formatter.go:58]: And this approach without goto is more easy to read/understand/more efficient?

We need to use runtime to complete this requirements - it was discussed at friday meeting.

**Review by rkiselenko — APPROVED** [review]
